src/provider/provider_coarse.c

```diff
@@ -0,0 +1,1431 @@
+err_delete_ravl_all_blocks:
+    ravl_delete(coarse_provider->all_blocks);
+err_delete_ravl_free_blocks:
+    ravl_delete(coarse_provider->free_blocks);
+err_delete_ravl_upstream_alloc:
+    ravl_delete(coarse_provider->upstream_alloc);
+err_free_coarse_provider:
+    umf_ba_global_free(coarse_provider);
+    return umf_result;
+}
+
+static void ravl_cb_rm_upstream_alloc_node(void *data, void *arg) {
+    assert(data);
+    assert(arg);
+
+    coarse_memory_provider_t *coarse_provider =
+        (struct coarse_memory_provider_t *)arg;
+    ravl_data_t *node_data = data;
+    block_t *alloc = node_data->value;
+    assert(alloc);
+
+    if (coarse_provider->upstream_memory_provider) {
+        // We continue to deallocate alloc blocks even if the upstream provider doesn't return success.
+        umfMemoryProviderFree(coarse_provider->upstream_memory_provider,
+                              alloc->data, alloc->size);
+    }
+
+    assert(coarse_provider->alloc_size >= alloc->size);
+    coarse_provider->alloc_size -= alloc->size;
+
+    umf_ba_global_free(alloc);
+}
+
+static void ravl_cb_rm_all_blocks_node(void *data, void *arg) {
+    assert(data);
+    assert(arg);
+
+    coarse_memory_provider_t *coarse_provider =
+        (struct coarse_memory_provider_t *)arg;
+    ravl_data_t *node_data = data;
+    block_t *block = node_data->value;
+    assert(block);
+
+    if (block->used) {
+        assert(coarse_provider->used_size >= block->size);
+        coarse_provider->used_size -= block->size;
+    }
+
+    if (block->free_list_ptr) {
+        free_blocks_rm_node(coarse_provider->free_blocks, block->free_list_ptr);
+    }
+
+    umf_ba_global_free(block);
+}
+
+static void coarse_memory_provider_finalize(void *provider) {
+    if (provider == NULL) {
+        assert(0);
+        return;
+    }
+
+    coarse_memory_provider_t *coarse_provider =
+        (struct coarse_memory_provider_t *)provider;
+
+    ravl_foreach(coarse_provider->all_blocks, ravl_cb_rm_all_blocks_node,
+                 coarse_provider);
+    assert(coarse_provider->used_size == 0);
+
+    ravl_foreach(coarse_provider->upstream_alloc,
+                 ravl_cb_rm_upstream_alloc_node, coarse_provider);
+    assert(coarse_provider->alloc_size == 0);
+
+    ravl_delete(coarse_provider->upstream_alloc);
+    ravl_delete(coarse_provider->all_blocks);
+    ravl_delete(coarse_provider->free_blocks);
+
+    umf_ba_global_free(coarse_provider);
+}
+
+static umf_result_t
+create_aligned_block(coarse_memory_provider_t *coarse_provider,
+                     size_t orig_size, size_t alignment, block_t **current) {
+    (void)orig_size; // unused in the Release version
+    int rv;
+
+    block_t *curr = *current;
+
+    // In case of non-zero alignment create an aligned block what would be further used.
+    uintptr_t original = (uintptr_t)curr->data;
+    uintptr_t aligned = (original + alignment - 1) & ~(alignment - 1);
```

Comment:
You could use `ALIGN_UP` macro directly instead.

Reply:
Done